icu4c/source/common/lsr.cpp

```diff
@@ -0,0 +1,118 @@
+
+// lsr.cpp
+// created: 2019may08 Markus W. Scherer
+
+#include "unicode/utypes.h"
+#include "cmemory.h"
+#include "cstring.h"
+#include "lsr.h"
+
+U_NAMESPACE_BEGIN
+
+LSR::LSR(char prefix, const char *lang, const char *scr, const char *r, UErrorCode &errorCode) :
+        language(nullptr), script(nullptr), region(r), owned(nullptr),
+        languageLength(0), scriptLength(0), regionIndex(indexForRegion(region)) {
+    if (U_SUCCESS(errorCode)) {
+        // Number of bytes for language + script with 2 prefixes and NUL terminators.
+        int32_t langSize = uprv_strlen(lang) + 1, scriptSize = uprv_strlen(scr) + 1;
+        int32_t langScriptSize = 2 + langSize + scriptSize;
+        char *p = owned = (char *)uprv_malloc(langScriptSize);
+        if (owned != nullptr) {
+            language = p;
+            *p++ = prefix;
+            uprv_memcpy(p, lang, langSize);
+            script = p += langSize;
+            *p++ = prefix;
+            uprv_memcpy(p, scr, scriptSize);
+            languageLength = langSize;
+            scriptLength = scriptSize;
+        } else {
+            errorCode = U_MEMORY_ALLOCATION_ERROR;
+        }
+    }
+}
+
+LSR::LSR(LSR &&other) U_NOEXCEPT :
+        language(other.language), script(other.script), region(other.region), owned(other.owned),
+        languageLength(other.languageLength), scriptLength(other.scriptLength),
+        regionIndex(other.regionIndex) {
+    if (owned != nullptr) {
+        other.language = other.script = "";
+        other.owned = nullptr;
+        other.languageLength = other.scriptLength = 0;
+    }
+}
+
+LSR::~LSR() {
+    uprv_free(owned);
+}
+
+LSR &LSR::operator=(LSR &&other) U_NOEXCEPT {
+    language = other.language;
+    script = other.script;
+    region = other.region;
+    languageLength = other.languageLength;
+    scriptLength = other.scriptLength;
+    regionIndex = other.regionIndex;
+    delete owned;
+    owned = other.owned;
+    if (owned != nullptr) {
+        other.language = other.script = "";
+        other.owned = nullptr;
+        other.languageLength = other.scriptLength = 0;
+    }
+    return *this;
+}
+
+UBool LSR::operator==(const LSR &other) const {
+    return languageLength == other.languageLength &&
+        uprv_memcmp(language, other.language, languageLength) == 0 &&
+        scriptLength == other.scriptLength &&
+        uprv_memcmp(script, other.script, scriptLength) == 0 &&
+        regionIndex == other.regionIndex &&
+        // Compare regions if both are ill-formed (and their indexes are 0).
+        (regionIndex > 0 || uprv_strcmp(region, other.region) == 0);
+}
+
+namespace {
+
+// Like uinvchar.h U_UPPER_ORDINAL(x) but with validation.
+// Returns 0..25 for A..Z else a value outside 0..25.
+inline int32_t upperOrdinal(int32_t c) {
+#if U_CHARSET_FAMILY==U_ASCII_FAMILY
+    return c - 'A';
+#elif U_CHARSET_FAMILY==U_EBCDIC_FAMILY
+    if (c <= 'I') { return c - 'A'; }
```

Comment:
Since you and I have no way to test that this works, I would prefer if you call a library function here.  Something more like,

```cpp
inline int32_t upperOrdinal(int32_t c) {
#if (non-ASCII)
  c = // (call library function to convert c to ASCII)
#endif
  return c - 'A';
}
```

Reply:
I prefer not to call a conversion function for this calculation.
Looking at an EBCDIC chart or a conversion table, this is not hard to follow.
I have also written this stuff before, see uinvchar.h/.cpp.

I will add a comments and a link.